src/otx/core/data/dataset/action_detection.py

```diff
@@ -0,0 +1,105 @@
+
+    def __init__(self, **kwargs) -> None:
+        super().__init__(**kwargs)
+        self.num_classes = len(self.dm_subset.categories()[AnnotationType.label])
+
+    def _get_item_impl(self, idx: int) -> ActionDetDataEntity | None:
+        item = self.dm_subset.get(id=self.ids[idx], subset=self.dm_subset.name)
+        img = item.media_as(Image)
+        img_data, img_shape = self._get_img_data_and_shape(img)
+
+        bbox_anns = [ann for ann in item.annotations if isinstance(ann, Bbox)]
+        bboxes = (
+            np.stack([ann.points for ann in bbox_anns], axis=0).astype(np.float32)
+            if len(bbox_anns) > 0
+            else np.zeros((0, 4), dtype=np.float32)
+        )
+
+        entity = ActionDetDataEntity(
+            image=img_data,
+            img_info=ImageInfo(
+                img_idx=idx,
+                img_shape=img_shape,
+                ori_shape=img_shape,
+                pad_shape=img_shape,
+                scale_factor=(1.0, 1.0),
+            ),
+            bboxes=tv_tensors.BoundingBoxes(
+                bboxes,
+                format=tv_tensors.BoundingBoxFormat.XYXY,
+                canvas_size=img_shape,
+            ),
+            labels=torch.nn.functional.one_hot(
+                torch.as_tensor([ann.label for ann in bbox_anns]),
+                self.num_classes,
+            ).to(torch.float),
+            frame_path=item.media.path,
+            proposals=self._get_proposals(
+                item.media.path,
+                self.dm_subset.infos().get(f"{self.dm_subset.name}_proposals", None),
+            ),
+        )
+
+        return self._apply_transforms(entity)
+
+    @staticmethod
+    def _get_proposals(frame_path: str, proposal_file: str | None) -> np.ndarray:
+        """Get proposal from frame path and proposal file name.
+
+        Datumaro AVA dataset expect data structure as
+        - data_root/
+            - frames/
+                - video0
+                    - video0_0001.jpg
+                    - vdieo0_0002.jpg
+            - annotations/
+                - train.csv
+                - val.csv
+                - train.pkl
+                - val.pkl
+        """
+        if proposal_file is None:
+            return np.array([[0, 0, 1, 1]], dtype=np.float64)
+
+        annotation_dir = Path(frame_path).parent.parent.parent
+        proposal_file_path = annotation_dir / "annotations" / proposal_file
+        if proposal_file_path.exists():
+            with Path.open(proposal_file_path, "rb") as f:
+                info = pickle.load(f)  # noqa: S301
+                if ",".join(Path(frame_path).stem.rsplit("_", 1)) in info:
+                    proposals = info[",".join(Path(frame_path).stem.rsplit("_", 1))][:, :4]
+                else:
+                    proposals = np.array([[0, 0, 1, 1]], dtype=np.float64)
+        else:
+            proposals = np.array([[0, 0, 1, 1]], dtype=np.float64)
+
+        return proposals
```

Comment:
```suggestion
        if not proposal_file_path.exists():
            return np.array([[0, 0, 1, 1]], dtype=np.float64)
        with Path.open(proposal_file_path, "rb") as f:
            info = pickle.load(f)  # noqa: S301

        return (
            info[",".join(Path(frame_path).stem.rsplit("_", 1))][:, :4]
            if ",".join(Path(frame_path).stem.rsplit("_", 1)) in info
            else np.array([[0, 0, 1, 1]], dtype=np.float64)
        )

        return proposals
```

Please reduce code nesting as possible for the better readability. In addition, why should weuse `np.float64` rather than `np.float32`?

Reply:
ddb54d7 Updated